esp-hal-common/src/interrupt/xtensa.rs

```diff
@@ -174,44 +179,350 @@ unsafe fn core1_interrupt_peripheral() -> *const crate::pac::interrupt_core1::Re
+        // Alternatively, a level-one interrupt can also be used instead (with software
+        // prioritization if needed) but I have no idea what software prioritization would
+        // look like in this context.
 
-#[no_mangle]
-#[link_section = ".rwtext"]
-fn __level_5_interrupt(_level: u32, save_frame: &mut Context) {
-    unsafe { level5_interrupt(save_frame) };
-}
+        // Priority4,
+        // Priority5,
+        // Priority6,
+    }
+
+    /// Stores what interrupts are enabled at each interrupt level
+    static mut INTERRUPT_LEVELS: [u128; 8] = [0u128; 8];
+
+    /// A priority of None, disables the interrupt
+    pub fn enable_with_priority(
+        core: crate::Cpu,
+        interrupt: Interrupt,
+        level: Priority,
+    ) -> Result<(), Error> {
+        let cpu_interrupt =
+            interrupt_level_to_cpu_interrupt(level, chip_specific::interrupt_is_edge(interrupt))?;
+        use xtensa_lx::mutex::Mutex as _;
+
+        #[cfg(any(feature = "esp32", feature = "esp32s3"))]
+        type Mutex<T> = xtensa_lx::mutex::CriticalSectionSpinLockMutex<T>;
+        #[cfg(feature = "esp32s2")]
+        type Mutex<T> = xtensa_lx::mutex::CriticalSectionMutex<T>;
+
+        // TODO: replace with critical-section CS later
+        static LEVELS_CS: Mutex<()> = Mutex::new(());
+        (&LEVELS_CS).lock(|_| unsafe {
+            // TODO reduce to 3?, as we only support up to 3 prios atm
+            for i in 0..=7 {
+                INTERRUPT_LEVELS[i] &= !(1 << interrupt.number());
+            }
+            INTERRUPT_LEVELS[level as usize] |= 1 << interrupt.number();
 
-#[no_mangle]
-#[link_section = ".rwtext"]
-fn __level_6_interrupt(_level: u32, save_frame: &mut Context) {
-    unsafe { level6_interrupt(save_frame) };
+            enable(core, interrupt, cpu_interrupt);
+
+            xtensa_lx::interrupt::enable_mask(
+                xtensa_lx::interrupt::get_mask() | 1 << cpu_interrupt as u32,
+            );
+        });
+        Ok(())
+    }
+
+    // TODO mention that theses interrupts are reservered in vector mode
+    // TODO make the normal `enable` unsafe? Would break vectoring unless careful
+    fn interrupt_level_to_cpu_interrupt(
+        level: Priority,
+        is_edge: bool,
+    ) -> Result<CpuInterrupt, Error> {
+        Ok(if is_edge {
+            match level {
+                Priority::None => return Err(Error::InvalidInterrupt),
+                Priority::Priority1 => CpuInterrupt::Interrupt10EdgePriority1,
+                Priority::Priority2 => return Err(Error::InvalidInterrupt),
+                Priority::Priority3 => CpuInterrupt::Interrupt22EdgePriority3,
+                // Priority::Priority4 => CpuInterrupt::Interrupt28EdgePriority4,
+                // Priority::Priority5 => CpuInterrupt::Interrupt31EdgePriority5,
+                // Priority::Priority6 => return Err(Error::InvalidInterrupt),
+                // Priority::Priority7 => CpuInterrupt::Interrupt14NmiPriority7,
```

Comment:
Can these just be removed? Same with the corresponding lines below.

Reply:
I think to remove these we would be making the executive decision to not support levels higher than 3 in Rust (freeing up levels 4-7 for asm usage). If we're all on board with that (I am) then lets do it.

Reply:
I agree on not supporting levels 4-7 in Rust 👍